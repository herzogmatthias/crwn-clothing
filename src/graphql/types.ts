import { ICartItem } from "../redux/cart/ICartItem";

export interface IHeaderType {
  cartHidden: boolean;
}

export interface ICacheState {
  cartHidden: boolean;
  cartItems: ICartItem[];
  itemCount: number;
}
