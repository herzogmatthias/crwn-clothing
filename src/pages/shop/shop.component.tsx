import * as React from "react";
import CollectionsOverview from "../../components/collections-overview/collections-overview.component";
import { Route, RouteComponentProps } from "react-router-dom";
import CollectionPage from "../collection/collection.component";
import {
  firestore,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";
import { updateCollections } from "../../redux/shop/shop.actions";
import { connect, ConnectedProps } from "react-redux";

type IShopPageProps = ConnectedProps<typeof connector> & RouteComponentProps;

function ShopPage({ match, updateCollections }: IShopPageProps) {
  let unsubscripeFromSnapshot = null;
  React.useEffect(() => {
    const collectionRef = firestore.collection("collections");
    collectionRef.onSnapshot(async (snapshot) => {
      const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
      updateCollections(collectionsMap);
    });
    return () => {};
  }, []);

  return (
    <div className="shop-page">
      <Route
        exact
        path={`${match.path}`}
        component={CollectionsOverview}
      ></Route>
      <Route
        path={`${match.path}/:categoryId`}
        component={CollectionPage}
      ></Route>
    </div>
  );
}

const mapDispatchToProps = {
  updateCollections: updateCollections,
};

const connector = connect(null, mapDispatchToProps);
export default connect(null, mapDispatchToProps)(ShopPage);
