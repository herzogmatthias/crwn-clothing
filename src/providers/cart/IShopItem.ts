export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
