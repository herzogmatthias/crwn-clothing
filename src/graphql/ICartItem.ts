import { IShopItem } from "./IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
