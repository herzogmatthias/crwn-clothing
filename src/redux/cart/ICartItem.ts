import { IShopItem } from "../shop/IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
