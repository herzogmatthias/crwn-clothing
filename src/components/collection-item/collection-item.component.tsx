import * as React from "react";
import { IShopItem } from "../../graphql/IShopItem";
import {
  ImageContainer,
  CollectionItemContainer,
  CollectionFooterContainer,
  NameContainer,
  PriceContainer,
  CustomButtonContainer,
} from "./collection-item.styles";
import { useMutation } from "react-apollo";
import { ADD_ITEM_TO_CART } from "../../graphql/mutations";

type ICollectionItemProps = {
  item: IShopItem;
  className?: string;
  addItem?(item: IShopItem): void;
};

function CollectionItem({ item, className }: ICollectionItemProps) {
  const [addItem] = useMutation<
    { additem(item: IShopItem): void },
    { item: IShopItem }
  >(ADD_ITEM_TO_CART);
  return (
    <CollectionItemContainer className={className}>
      <ImageContainer imageUrl={item.imageUrl}></ImageContainer>
      <CollectionFooterContainer>
        <NameContainer>{item.name}</NameContainer>
        <PriceContainer>${item.price}</PriceContainer>
      </CollectionFooterContainer>
      <CustomButtonContainer
        onClick={() => addItem({ variables: { item } })}
        inverted
      >
        Add to cart
      </CustomButtonContainer>
    </CollectionItemContainer>
  );
}

export default CollectionItem;
