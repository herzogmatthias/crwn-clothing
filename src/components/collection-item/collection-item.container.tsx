import React from "react";
import { IShopItem } from "../../redux/shop/IShopItem";
import { Mutation } from "react-apollo";
import { ADD_ITEM_TO_CART } from "../../graphql/mutations";
import CollectionItem from "./collection-item.component";

interface IContainerProps {
  item: IShopItem;
}

const CollectionItemContainer = (props: IContainerProps) => {
  return (
    <Mutation<{}, { item: IShopItem }> mutation={ADD_ITEM_TO_CART}>
      {(addItemToCart) => (
        <CollectionItem
          {...props}
          addItem={(item) => addItemToCart({ variables: { item } })}
        />
      )}
    </Mutation>
  );
};

export default CollectionItemContainer;
