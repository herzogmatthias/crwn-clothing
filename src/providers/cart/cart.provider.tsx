import React, { createContext, useEffect } from "react";
import { ICartProvider } from "./ICartProvider";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  addItemToCart,
  removeItemFromCart,
  filterItemFromCart,
  getCartItemsCount,
  getTotalPrice,
} from "../../redux/cart/cart.utils";
import { IShopItem } from "../../redux/shop/IShopItem";

export const CartContext = createContext<ICartProvider>({
  hidden: true,
  total: 0,
  cartItems: [],
  cartItemsCount: 0,
  toggleHidden: () => {},
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
});

interface ICartProviderProps {
  children: React.ReactNode;
}

const CartProvider = ({ children }: ICartProviderProps) => {
  const [hidden, setHidden] = React.useState(true);

  const [cartItems, setCartItems] = React.useState<ICartItem[]>([]);
  const [cartItemsCount, setCartItemsCount] = React.useState(0);

  const [total, setTotal] = React.useState(0);

  const toggleHidden = () => {
    setHidden(!hidden);
  };

  const addItem = (item: IShopItem) =>
    setCartItems(addItemToCart(cartItems, item));

  const removeItem = (item: ICartItem) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const clearItemFromCart = (item: ICartItem) =>
    setCartItems(filterItemFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setTotal(getTotalPrice(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        total,
        addItem,
        hidden,
        toggleHidden,
        cartItems,
        cartItemsCount,
        removeItem,
        clearItemFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
