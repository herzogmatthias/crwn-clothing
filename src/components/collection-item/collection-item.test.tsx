import React from "react";
import { IShopItem } from "../../redux/shop/IShopItem";
import { ShallowWrapper, shallow } from "enzyme";
import { ActionCreatorWithPayload } from "@reduxjs/toolkit";
import { ICartItem } from "../../redux/cart/ICartItem";
import { ADD_ITEM } from "../../redux/cart/cart.types";
import { CollectionItem } from "./collection-item.component";
import {
  CustomButtonContainer,
  NameContainer,
  PriceContainer,
  ImageContainer,
} from "./collection-item.styles";
describe("NON CONNECTED COLLECTION-ITEM TESTS", () => {
  let wrapper: ShallowWrapper;
  let mockAddItem: ActionCreatorWithPayload<IShopItem, "ADD_ITEM">;
  const dispatch = jest.fn();
  const item: IShopItem = {
    id: 2,
    imageUrl: "asdf234",
    name: "hat",
    price: 34,
  };
  beforeEach(() => {
    mockAddItem = jest.fn((item: ICartItem) => {
      dispatch({ type: ADD_ITEM, payload: item });
    }) as any;
    const mockProps = {
      item,
      addItem: mockAddItem,
    };
    wrapper = shallow(<CollectionItem {...mockProps} />);
  });
  test("render CollectionItem Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("call addItem when AddButton clicked", () => {
    wrapper.find(CustomButtonContainer).simulate("click");

    expect(mockAddItem).toHaveBeenCalledWith(item);
  });

  test("render imageUrl as a prop on BackgroundImage", () => {
    expect(wrapper.find(ImageContainer).prop("imageUrl")).toBe(item.imageUrl);
  });

  test("render name prop in NameContainer", () => {
    expect(wrapper.find(NameContainer).text()).toBe(item.name);
  });

  test("render price prop in PriceContainer", () => {
    const price = parseInt(wrapper.find(PriceContainer).text().substr(1, 2));
    expect(price).toBe(item.price);
  });
});
