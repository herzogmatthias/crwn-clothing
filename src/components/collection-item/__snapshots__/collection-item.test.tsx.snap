// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NON CONNECTED COLLECTION-ITEM TESTS render CollectionItem Component 1`] = `
<styled.div>
  <styled.div
    imageUrl="asdf234"
  />
  <styled.div>
    <styled.span>
      hat
    </styled.span>
    <styled.span>
      $
      34
    </styled.span>
  </styled.div>
  <Styled(CustomButton)
    inverted={true}
    onClick={[Function]}
  >
    Add to cart
  </Styled(CustomButton)>
</styled.div>
`;
